src/analysis/calculus/specific_functions.lean

```diff
@@ -185,3 +186,143 @@ begin
+(div_eq_one_iff_eq $ (pos_denom x).ne').2 $ by rw [zero_of_nonpos (sub_nonpos.2 h), add_zero]
+
+lemma zero_of_nonpos (h : x ≤ 0) : smooth_transition x = 0 :=
+by rw [smooth_transition, zero_of_nonpos h, zero_div]
+
+lemma le_one (x : ℝ) : smooth_transition x ≤ 1 :=
+(div_le_one (pos_denom x)).2 $ le_add_of_nonneg_right (nonneg _)
+
+lemma nonneg (x : ℝ) : 0 ≤ smooth_transition x :=
+div_nonneg (exp_neg_inv_glue.nonneg _) (pos_denom x).le
+
+lemma lt_one_of_lt_one (h : x < 1) : smooth_transition x < 1 :=
+(div_lt_one $ pos_denom x).2 $ lt_add_of_pos_right _ $ pos_of_pos $ sub_pos.2 h
+
+lemma pos_of_pos (h : 0 < x) : 0 < smooth_transition x :=
+div_pos (exp_neg_inv_glue.pos_of_pos h) (pos_denom x)
+
+protected lemma times_cont_diff : times_cont_diff ℝ ⊤ smooth_transition :=
+smooth.div (smooth.add $ smooth.comp $ times_cont_diff_const.sub times_cont_diff_id) $
+  λ x, (pos_denom x).ne'
+
+protected lemma times_cont_diff_at {x n} : times_cont_diff_at ℝ n smooth_transition x :=
+(smooth_transition.times_cont_diff.of_le le_top).times_cont_diff_at
+
+end smooth_transition
+
+variables {E F : Type*} [inner_product_space ℝ E] [normed_group F] [normed_space ℝ F]
+
+lemma times_cont_diff_inner {n} : times_cont_diff ℝ n (λ p : E × E, ⟪p.1, p.2⟫_ℝ) :=
+is_bounded_bilinear_map.times_cont_diff
+{ add_left := λ _ _ _, inner_add_left,
+  smul_left := λ _ _ _, inner_smul_left,
+  add_right := λ _ _ _, inner_add_right,
+  smul_right := λ _ _ _, inner_smul_right,
+  bound := ⟨1, zero_lt_one, λ x y, by { rw one_mul, exact abs_real_inner_le_norm x y, }⟩ }
+
+lemma times_cont_diff_at_inner {p : E × E} {n} :
+  times_cont_diff_at ℝ n (λ p : E × E, ⟪p.1, p.2⟫_ℝ) p :=
+times_cont_diff_inner.times_cont_diff_at
+
+lemma times_cont_diff_within_at.inner {f g : F → E} {s : set F} {x : F} {n : with_top ℕ}
+  (hf : times_cont_diff_within_at ℝ n f s x) (hg : times_cont_diff_within_at ℝ n g s x) :
+  times_cont_diff_within_at ℝ n (λ x, ⟪f x, g x⟫_ℝ) s x :=
+times_cont_diff_at_inner.comp_times_cont_diff_within_at x (hf.prod hg)
+
+lemma times_cont_diff_at.inner {f g : F → E} {x : F} {n : with_top ℕ}
+  (hf : times_cont_diff_at ℝ n f x) (hg : times_cont_diff_at ℝ n g x) :
+  times_cont_diff_at ℝ n (λ x, ⟪f x, g x⟫_ℝ) x :=
+hf.inner hg
+
+lemma real.times_cont_diff_at_sqrt {x : ℝ} (hx : x ≠ 0) {n} :
+  times_cont_diff_at ℝ n real.sqrt x :=
+sorry
+
+lemma times_cont_diff_at_norm {x : E} (hx : x ≠ 0) {n} : times_cont_diff_at ℝ n norm x :=
+begin
+  rw [show norm = λ x : E, real.sqrt ⟪x, x⟫_ℝ, from funext norm_eq_sqrt_real_inner],
+  exact (real.times_cont_diff_at_sqrt (mt inner_self_eq_zero.1 hx)).comp x
+    (times_cont_diff_at_id.inner times_cont_diff_at_id)
+end
+
+/-- A function `f : E → ℝ` defined on a real inner product space with the following properties:
+
+- `f` is infinitely smooth on `E`;
+- `f` is positive on `ball 0 2` and equals zero otherwise;
+- `f` is equal to `1` on `closed_ball 0 1`. -/
+def smooth_bump_function (x : E) :=
+smooth_transition (2 - ∥x∥)
```

Comment:
If you use `(4 - ∥x∥ ^ 2) / 3` instead of `2 - ∥x∥`, you can get this without the results of #4407, I think.

Reply:
You're right, thanks!

Reply:
I've decided to stick to the original plan.